### DISCUSSION TITLE
Working in python from R-Instat?

### DISCUSSION THREAD

**Discussion by rdstern** [IDEMSInternational/R-Instat#8045] [Ideas]
@AmsaleEjigu and @lloyddewit I wonder about the practicalities about python from R-Instat - possibly just from the (new) script window?  And, if that is interesting, then should we be looking at the[ reticulate](https://cran.r-project.org/web/packages/reticulate/index.html) package?  It says it is:

```
Interface to 'Python'
Interface to 'Python' modules, classes, and functions. When calling into 'Python', R data types are automatically converted to their equivalent 'Python' types. When values are returned from 'Python' to R they are converted back to R types. Compatible with all versions of 'Python' >= 2.7.
```
It includes a whole set of vignettes of which one is called:

[Primer on Python for R Users](https://cran.r-project.org/web/packages/reticulate/vignettes/python_primer.html)

One context I am thinking of is the  @jkmusyoka MSc course on Data Science for Development, which should be starting in September 2023.  This will use R-Instat and (later) there will probably be a course on R programming, and another on python programming.  They may be optional courses.  Just maybe there should be a single course on both languages?

**Comment by lloyddewit** [comment]
@rdstern I remembered that we also discussed this in July 2022. For info, I added the email thread below (email addresses redacted).

----

**Sent: Thursday, 28 July 2022 09:26
Subject: RE: Combined file for R-Instat using python**

Roger, Amsale

Thank you for the responses. You have the best understanding of the user needs and I’m happy with whatever decisions you take. My role is to point out the technical risks.

Roger’s suggestion to first add this as an experimental feature sounds sensible.

Stephen

**Sent: Thursday, 28 July 2022 09:18
Subject: Re: Combined file for R-Instat using python**

Hi Stephen and Roger,

Thank you for your replies.

In this particular case, even rewriting it in R will not be that difficult. However, it would be great if we can have the facility to run Python scripts in R-Instat in the future. There are many people, in AIMS for example, who can be happy to continue using Python and still use R-Instat.

Kind regards
Amsale

**Sent: Thursday, 28 July 2022 09:15
Subject: RE: Combined file for R-Instat using python**

Hi all,

At this stage I am just interested in investigating how easily this facility can be added to R-Instat.  I am not assuming it will be used by many R-Instat users, so (if it works) it may remain simply as a series of instructions on how this could be made possible for those who may need it.  I am not assuming the ordinary installation would include it at all.

From the review it was clear that Bob anticipates that many users of R-Instat (and other GUIs) will not be happy at all with commands.  They may then need help occasionally from experienced computer people on tricky problems that are not possible, or inefficient to do from the menus.  We had one such problem on this workshop.  It was as follows:

a)	There were 57 files that needed to be read into R-Instat and reorganised into a single data frame.
b)	We could do this quite easily – and slowly – in R-Instat.
c)	Meanwhile the producer of the problem tried with 1000 files and it seemed to take forever.
d)	And he said he had about 500,000 of these files to merge!
e)	So on these occasions he needs help from a guru.  Hopefully an R-guru.  
f)	But really nice, if they found a python guru instead, that this could be accommodated – perhaps with some pain.
g)	And one clear audience for this is the AIMS students, for whom we teach 2 courses using R-Instat.  And they are all also taught python!  And some love it!

Cheers

Roger

**On Thu, 28 Jul 2022 at 07:50**
Roger, Amsale
 
Thank you for the mail.
 
I understand that you are interested in running Python scripts from the R-Instat script window. If we add this feature, then we increase R-Instat’s dependencies and complexities. We may need to update the build/installation and ensure that there is a compatible Python environment installed on the user’s machine. It would require extra regression testing. Also, I’ve been working a lot recently with calling R from Python. It is not as simple as it appears and there are many complexities and pitfalls. Therefore, I think we should only add this feature if we think it will be widely used. 
 
I would personally first try to do the merging in R. Or alternatively use a script that can be called from outside R-Instat. 
 
Hope this helps
 
Stephen
 
 **Sent: Thursday, 28 July 2022 07:22
Subject: RE: Combined file for R-Instat using python**
 
Hi Amsale,
 
That sounds great.  I note there is the R package, called reticulate, that allows a python program to be run from R?  I would really like us to be able to supply an r-script that they could use themselves – and not depend on us.
 
Could you try adding that package to R-Instat and see if we could use your Python script, through that, in an R-Instat script.  
 
I am copying Stephen on this message.  He is improving the facilities in an R-Instat script window and, if this doesn’t work yet, then perhaps this item could be included in his topics for upgrading the running of R-scripts in R-Instat.
 
An alternative would be to translate your python script into R.  That could be good to ask Lily about, but if we could run python scripts from R-Instat that would be a really great feature to have.  So let’s start by investigating your python route!
 
Cheers
 
Roger
